core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/SSHAction.java

```diff
@@ -0,0 +1,150 @@
+ * A user must specify username and keypair authentication credentials.
+ * Optionals include port and machine URL
+ */
+@Plugin(type = Action.PLUGIN_TYPE)
+@Name("SSHAction")
+@Description("Action to run a script on remote machine")
+public class SSHAction extends Action {
+  private final SSHActionConfig config;
+
+  private static final Logger LOG = LoggerFactory.getLogger(SSHAction.class);
+
+  public SSHAction(SSHActionConfig config) {
+    this.config = config;
+  }
+
+  @Override
+  public void run(final ActionContext context) throws Exception {
+    Connection connection = new Connection(config.host);
+    Session session = connection.openSession();
+    try {
+      connection.connect();
+
+      String password = Strings.isNullOrEmpty(config.password) ? null : config.password;
+      if (!connection.authenticateWithPublicKey(config.user, new File(config.privateKeyFile), password)) {
+        throw new IOException(String.format("Unable to establish SSH connection for %s@%s on port %d",
+                                            config.user, config.host, config.port));
+      }
+
+      session.execCommand(config.cmd);
+
+      InputStream stdout = new StreamGobbler(session.getStdout());
+      BufferedReader outBuffer = new BufferedReader(new InputStreamReader(stdout, Charsets.UTF_8));
+      InputStream stderr = new StreamGobbler(session.getStderr());
+      BufferedReader errBuffer = new BufferedReader(new InputStreamReader(stderr, Charsets.UTF_8));
+
+      String out = CharStreams.toString(outBuffer);
+      String err = CharStreams.toString(errBuffer);
+
+      LOG.info("Errors: {}", err);
+
+      if (config.output != null) {
+        context.getArguments().set(config.output, out);
+      }
+    } finally {
+      connection.close();
+      session.close();
+    }
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    this.config.validate();
+  }
+
+  /**
+   * Config class that contains all the properties needed to SSH into the remote machine and run the script.
+   */
+  public static class SSHActionConfig extends PluginConfig {
+
+    @Description("Host name of the remote machine where the command needs to be executed.")
+    private String host;
+
+    @Nullable
+    @Description("Port to connect to. Defaults to 22")
+    private Integer port;
+
+    @Description("User name used to connect to host")
+    private String user;
+
+    @Description("File path to Private key")
+    private String privateKeyFile;
+
+    @Description("Password associated with private key")
+    private String password;
+
+    @Description("Command to be executed on the remote host.")
+    private String cmd;
+
+    @Nullable
+    @Description("Script output variable")
+    private String output;
+
+    public SSHActionConfig() {
+      port = 22;
+    }
+
+    public SSHActionConfig(String host, String user, String privateKeyFile, int port, String password,
```

Comment:
yes. Currently I didn't include unit tests, because we're trying to figure out if we can even do unit tests. LMK if I should remove this.